storage-proofs/src/hasher/pedersen.rs

```diff
@@ -249,6 +306,96 @@ impl LightAlgorithm<PedersenDomain> for PedersenFunction {
     }
 }
 
+impl PedersenFunction {
+    pub fn new() -> Self {
+        PedersenFunction::default()
+    }
+
+    /// Allows the user to use `PedersenFunction` as an updatable hasher where the user passes in
+    /// data continuously as opposed to all at once. This method bitwise concatenates the `data`
+    /// argument with any previously passed in data and updates the stored hash `self.curr_hash` if
+    /// more than one Pedersen segment's worth of data is available.
+    ///
+    /// When all preimage data has been passed in via one or more calls to `self.update()`, the user
+    /// must call `self.finalize()` to retrieve the hash digest.
+    pub fn update(&mut self, data: &[u8]) {
+        let input_bits = BitVec::<LittleEndian, u8>::from(data);
+
+        let n_input_bits = input_bits.len();
+        let n_stored_bits = self.unhashed_bits.len();
+        let n_bits_total = n_input_bits + n_stored_bits;
+
+        if n_bits_total < N_SEGMENT_BITS {
+            self.unhashed_bits.extend(input_bits);
+            return;
+        }
+
+        let n_bits_to_store = n_bits_total % N_SEGMENT_BITS;
+        let n_bits_to_hash = n_bits_total - n_bits_to_store;
+
+        let bits_to_hash = self
```

Comment:
I had a manual implementation of this chaining before, as this is pretty slow in bitvec if I remember correctly :/

Reply:
see `NodeBits` below